### DISCUSSION TITLE
How to package my code with all metadata included？

### DISCUSSION THREAD

**Discussion by lzt007** [pyinstaller/pyinstaller#8262] [Help] [answered]
I need to package my code with some packages used by side-modules which are not directly called by main-code.
For example, a side-module named ModuleA.py using package "requests".As requests was not included in the main-code，I use
`--hidden-import=ModuleA` to make the package "requests" included into the program.
By this way, the metadata of "requests" will not be included, so does the metadata of the packages needed by "requests".That's ok for running the program but for some reason, I want's to collect all the metadatas of all the packages that were included by all my modules and instert them into my program.
I have read the doc and tried the params like:
`--collect-all=requests`
`--recursive-copy-metadata=requests`
These params work well and the metadata of requests was truely included.But when I try:
`--collect-all=ModuleA`
`--recursive-copy-metadata=ModuleA`
For `--collect-all` , it looks like `--collect-all` treat packages imported by the module as `--hidden-import` and won‘t copy the metadata of them. And for `--recursive-copy-metadata`, according to the doc, that param can't work with modulename and just raise some error.

I have many Modules like ModuleA, with numbers of packages included in them,it's a disaster to manually find all the packages(especally the packages used by the packages) and insert them by params one by one.So is there a way to get all the metadatas included by module name rather than package name?

**Comment by rokm** [comment]
> So is there a way to get all the metadatas included by module name rather than package name?

There is not.

But actually, what is a `ModuleA`, and where is it located? Is it a module (`ModuleA.py`), an importable package, or a `pip`-installed dist (which provides both importable package and metadata)?

**Reply by lzt007** [reply]
ModuleA means a file like ModuleA.py or a folder named ModuleA with __init__.py. 
It‘s not a third-part package installed by pip but may has some pip-packages imported in it.

So the only way to package all the metadatas is adding `--recursive-copy-metadata=every package imported in Module` one by one?

**Reply by rokm** [reply]
So it doesn't have metadata at all, and that's why ` --recursive-copy-metadata=ModuleA` does nothing.

**Reply by lzt007** [reply]
So if I write a metadata for ModuleA, that may work well?

**Reply by bwoodsend** [reply, marked as answer]
If you turn ModuleA into a proper, [pip installable distribution](https://packaging.python.org/en/latest/tutorials/packaging-projects/), with ModuleA's dependencies [declared as such](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#dependencies-and-requirements) then `--recursive-copy-metadata=ModuleA` will do what you're after.

**Reply by lzt007** [reply]
well，that sounds not much better than add packages one by one. thanks anyway.
